README.md
# Poly-REaDDIT
## A tool for gathering information and Polygon address NFT details about any Reddit users.
Since last year Reddit introduced a way to costumize your profile with NFTs. This tool is a way to gather information, its NFTs and the Polygon address of any Reddit user.

<div align="center">
     <img alt="Avatar" src="img/img.png">
</div>

When you purchase or mint a Reddit NFT, Reddit client will create a Polygon wallet/address for you. This address is linked to your Reddit account and can be used to send and receive NFTs.
The wallet is related to the Poligon network, so you can use it to send and receive NFTs from other Polygon wallets or sell it. Poligon is a Layer 2 scaling solution for Ethereum.

The NFT will bring the ability to create a new revenue stream for Reddit and its users, and also customize the aspect of the profile. The related NFTs are listed on OpenSea and PolygonScan.



## Installation and usage
```
$ git clone
$ cd Poly-REaDDIT
$ pip install -r requirements.txt
```
Or directly install the requirements with pip:
```
$ pip install poly-readdit
```
Then you can run the tool with:
```
$ python ./poly-readdit/poly-readdit.py -username xxxxxx
```
Or if you installed via pip directly with:
```
$ polyreaddit -username xxxxxx
```
## Example
```
./poly-readdit/poly-readdit.py -username xxxxxxx

⠀⠀⠀⢠⣾⣿⣿⣿⣿⣶⣤⣤⣾⠛⠻⣷⡀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⣿⣿⡏⠉⠉⠙⠛⠿⠿⣷⣀⣀⣿⠃⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⡇⠀⠀⠀⠀⠀⠀⠈⠉⠉⠁⠀⠀
⠀⠀⣀⣤⣀⠀⢀⣠⣤⣶⣶⣿⣿⣿⣿⣿⣿⣶⣶⣤⣄⡀⠀⣀⣤⣀⠀⠀
⢰⡿⠋⢉⣹⣿⣿⣿⠿⠟⠛⠋⠉⠉⠉⠉⠙⠛⠻⠿⣿⣿⣿⣏⡉⠙⢿⡆
⢸⣇⣠⣾⣿⡿⠋⠀⠀⣠⣤⣀⠀⠀⠀⠀⣀⣤⣄⠀⠀⠙⢿⣿⣷⣄⣸⡗
⠈⢻⣿⣿⠋⠀⠀⠀⢸⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⡇⠀⠀⠀⠙⣿⣿⡟⠁
⠀⢸⣿⣿⠀⠀⠀⠀⠀⠉⠋⠁⠀⠀⠀⠀⠈⠙⠉⠀⠀⠀⠀⠀⣿⣿⡇⠀
⠀⠀⣿⣿⣧⡀⠀⠀⠀⢤⣀⡀⠀⠀⠀⠀⢀⣀⡤⠀⠀⠀⢀⣼⣿⣿⠀⠀
⠀⠀⠈⠿⣿⣷⣦⣀⠀⠀⠉⠻⠿⠿⠿⠿⠟⠉⠀⠀⣀⣴⣾⣿⠿⠁⠀⠀
⠀⠀⠀⠀⠉⠻⢿⣿⣿⣷⣶⣤⣤⣤⣤⣤⣤⣶⣾⣿⣿⡿⠟⠉⠀   Poly-REaDDIT: ⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⠉⠛⠛⠻⠿⠿⠿⠿⠟⠛⠛⠉⠁⠀⠀⠀⠀   a Reddit user info collector⠀⠀⠀


 [+] REDDIT USER INFO
  |
  ├ Username: xxxx
  ├ Created UTC: 2013-10-31 09:43:58
  ├ User ID: t2_xxx
  ├ Is mod: False
  ├ Is NSFW: False
  ├ Is Gold: False
  ├ Is Employee: False
  ├ Is Suspended: False
  ├ Hide from robots: True
  ├ Profile ID: t5_xxx
  ├ User name: u/axxxxx
  ├ User karma: 3 (0: Awards received, 0: Awards given, 1: Comments, 2: Posts)
  ├ Polygon Address: 0x5exxxxxxx
  └ ------------------------------------

 [+] NFT INFO
  |
  ├ Name: Hog-nosed Bats #71xxxxx
  ├ ID: nft_eip155:137_f33ad86bb54a2xxxx
  ├ Serial Number: 71xxxx
  ├ Title: Hog-nosed Bats #71xxxx
  ├ External Urls: https://polygonscan.com/token/0xf33ad86bb54xxxxx
  └ ------------------------------------

  ├ Name: The Big Bucks #1148
  ├ ID: nft_eip155:137_4bexxxxxx
  ├ Serial Number: 1xxx
  ├ Title: XXXXXXXXX #1xxxx
  ├ External Urls: https://polygonscan.com/token/0x4be1e0xxxxxxx
  └ ------------------------------------
```

CHANGES.md
# CHANGELOG

## VER 0.0.1
- Create the tool
- Introduced command line utils `poly-readdit`

main.py
import argparse
import requests
import json
from bs4 import BeautifulSoup
from datetime import datetime


class PolyReaddit:

    start_headers = {
        "Accept-Language": "it-IT,it;q=0.8,en-US;q=0.5,en;q=0.3",
        "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/111.0",
    }
    session = None
    session_api = None

    username = ""
    created_str = ""
    nfts = []
    user_id = ""
    is_mod = ""
    is_NSFW = ""
    is_Gold = ""
    is_Employee = ""
    is_Suspended = ""
    hide_from_robots = ""
    profile_id = "N/A"
    user_name = ""
    user_karma = ""
    reddit_access_token = ""
    polygon_address = ""

    def print_banner(self):
        print(
            """
⠀⠀⠀⢠⣾⣿⣿⣿⣿⣶⣤⣤⣾⠛⠻⣷⡀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⣿⣿⡏⠉⠉⠙⠛⠿⠿⣷⣀⣀⣿⠃⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⡇⠀⠀⠀⠀⠀⠀⠈⠉⠉⠁⠀⠀
⠀⠀⣀⣤⣀⠀⢀⣠⣤⣶⣶⣿⣿⣿⣿⣿⣿⣶⣶⣤⣄⡀⠀⣀⣤⣀⠀⠀
⢰⡿⠋⢉⣹⣿⣿⣿⠿⠟⠛⠋⠉⠉⠉⠉⠙⠛⠻⠿⣿⣿⣿⣏⡉⠙⢿⡆
⢸⣇⣠⣾⣿⡿⠋⠀⠀⣠⣤⣀⠀⠀⠀⠀⣀⣤⣄⠀⠀⠙⢿⣿⣷⣄⣸⡗
⠈⢻⣿⣿⠋⠀⠀⠀⢸⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⡇⠀⠀⠀⠙⣿⣿⡟⠁
⠀⢸⣿⣿⠀⠀⠀⠀⠀⠉⠋⠁⠀⠀⠀⠀⠈⠙⠉⠀⠀⠀⠀⠀⣿⣿⡇⠀
⠀⠀⣿⣿⣧⡀⠀⠀⠀⢤⣀⡀⠀⠀⠀⠀⢀⣀⡤⠀⠀⠀⢀⣼⣿⣿⠀⠀
⠀⠀⠈⠿⣿⣷⣦⣀⠀⠀⠉⠻⠿⠿⠿⠿⠟⠉⠀⠀⣀⣴⣾⣿⠿⠁⠀⠀
⠀⠀⠀⠀⠉⠻⢿⣿⣿⣷⣶⣤⣤⣤⣤⣤⣤⣶⣾⣿⣿⡿⠟⠉⠀   Poly-REaDDIT: ⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⠉⠛⠛⠻⠿⠿⠿⠿⠟⠛⠛⠉⠁⠀⠀⠀⠀   a Reddit user info collector⠀⠀⠀
"""
        )

    def __init__(self, username):
        self.print_banner()
        self.reddit_session = requests.Session()
        self.reddit_session.headers = self.start_headers
        self.username = username
        self.get_reddit_user_info()
        self.get_NFT()

    def get_reddit_user_info(self):
        try:
            url = "https://www.reddit.com/user/" + self.username
            r = self.reddit_session.get(url)
            soup = BeautifulSoup(r.text, "html.parser")
            data = soup.find("script", {"id": "data"}).text
            body_data = json.loads(data.split(" = ")[1][:-1])
            created = datetime.utcfromtimestamp(
                body_data["users"]["models"][self.username.lower()]["createdUtc"]
            )
            self.created_str = created.strftime("%Y-%m-%d %H:%M:%S")
            self.user_id = body_data["users"]["models"][self.username.lower()]["id"]
            self.is_mod = body_data["users"]["models"][self.username.lower()]["isMod"]
            self.is_NSFW = body_data["users"]["models"][self.username.lower()]["isNSFW"]
            self.is_Gold = body_data["users"]["models"][self.username.lower()]["isGold"]
            self.is_Employee = body_data["users"]["models"][self.username.lower()][
                "isEmployee"
            ]
            self.is_Suspended = body_data["users"]["models"][self.username.lower()][
                "isSuspended"
            ]
            self.hide_from_robots = body_data["users"]["models"][self.username.lower()][
                "hideFromRobots"
            ]
            self.user_name = body_data["users"]["models"][self.username.lower()][
                "displayNamePrefixed"
            ]
            if body_data["users"]["models"][self.username.lower()]["hasUserProfile"]:
                self.profile_id = body_data["users"]["models"][self.username.lower()][
                    "profileId"
                ]
            self.user_karma = body_data["users"]["models"][self.username.lower()][
                "karma"
            ]["total"]
            self.user_karma_award_rec = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromAwardsReceived"]
            self.user_karma_award_given = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromAwardsGiven"]
            self.user_karma_comments = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromComments"]
            self.user_karma_post = body_data["users"]["models"][self.username.lower()][
                "karma"
            ]["fromPosts"]
            self.reddit_access_token = body_data["user"]["session"]["accessToken"]

        except:
            print(" [-] Error getting user info")
            return False

    def get_NFT(self):
        try:
            url = "https://gql.reddit.com/"
            self.session_api = requests.Session()
            _head = self.start_headers
            _head["Authorization"] = "Bearer " + self.reddit_access_token.replace(
                "'", ""
            )
            _head["Referer"] = "https://www.reddit.com/"
            _head["Origin"] = "https://www.reddit.com"
            _head["Content-Type"] = "application/json"
            self.session_api.headers = _head
            _payload = {
                "variables": {"id": self.user_id, "isLoggedInUserProfile": False},
                "id": "985a7e8ad0b9",
            }
            r = self.session_api.post(url, json=_payload)
            json_data = json.loads(r.text)
            for i in json_data["data"]["redditorInfoById"]["displayedCollectibleItems"][
                "edges"
            ]:
                nft = {}
                nft["name"] = i["node"]["item"]["name"]
                nft["id"] = i["node"]["item"]["id"]
                nft["serialNumber"] = i["node"]["item"]["serialNumber"]
                nft["title"] = i["node"]["item"]["nft"]["title"]
                nft["externalUrls"] = i["node"]["item"]["nft"]["externalUrls"][0]
                self.nfts.append(nft)
                if self.polygon_address == "":
                    self.get_token_owner(i["node"]["item"]["nft"]["externalUrls"][0])
        except:
            print(" [-] Error getting NFT")

    def print_data(self):
        print("")
        print(" [+] REDDIT USER INFO")
        print(" [|] Username: {}".format(self.username))
        print(" [|] Created UTC: {}".format(self.created_str))
        print(" [|] User ID: {}".format(self.user_id))
        print(" [|] Is mod: {}".format(self.is_mod))
        print(" [|] Is NSFW: {}".format(self.is_NSFW))
        print(" [|] Is Gold: {}".format(self.is_Gold))
        print(" [|] Is Employee: {}".format(self.is_Employee))
        print(" [|] Is Suspended: {}".format(self.is_Suspended))
        print(" [|] Hide from robots: {}".format(self.hide_from_robots))
        print(" [|] Profile ID: {}".format(self.profile_id))
        print(" [|] User name: {}".format(self.user_name))
        print(
            " [|] User karma: {} ({}: Awards received, {}: Awards given, {}: Comments, {}: Posts)".format(
                self.user_karma,
                self.user_karma_award_rec,
                self.user_karma_award_given,
                self.user_karma_comments,
                self.user_karma_post,
            )
        )
        print(" [|] Polygon Address: {}".format(self.polygon_address))
        print("")

        print(" [+] NFT INFO")
        for i in self.nfts:
            print(" [|] Name: {}".format(i["name"]))
            print(" [|] ID: {}".format(i["id"]))
            print(" [|] Serial Number: {}".format(i["serialNumber"]))
            print(" [|] Title: {}".format(i["title"]))
            print(" [|] External Urls: {}".format(i["externalUrls"]))
            print("")

    def get_token_owner(self, url):
        try:
            _query = url.split("token/")[1].split("?a=")
            _token = _query[0]
            _id = _query[1]
            text = requests.get(
                "https://opensea.io/assets/matic/{}/{}".format(_token, _id),
                headers=self.start_headers,
            ).text
            soup = BeautifulSoup(text, "html.parser")
            link = soup.find(
                "div", {"data-testid": "ItemOwnerAccountLink"}
            ).findChildren("a")
            self.polygon_address = link[0].attrs["href"].replace("/", "")
        except:
            print(" [-] Error getting token owner")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-username", help="reddit username")
    args = parser.parse_args()
    poly = PolyReaddit(args.username)
    poly.print_data()


src/poly-readdit/poly-readdit.py
import argparse
import requests
import json
from bs4 import BeautifulSoup
from datetime import datetime


class PolyReaddit:

    start_headers = {
        "Accept-Language": "us-US,us;q=0.8,en-US;q=0.5,en;q=0.3",
        "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/111.0",
    }
    session = None
    session_api = None

    username = ""
    created_str = ""
    nfts = []
    user_id = ""
    is_mod = ""
    is_NSFW = ""
    is_Gold = ""
    is_Employee = ""
    is_Suspended = ""
    hide_from_robots = ""
    profile_id = "N/A"
    user_name = ""
    user_karma = ""
    reddit_access_token = ""
    polygon_address = "N/A"

    def print_banner(self):
        print(
            """
⠀⠀⠀⢠⣾⣿⣿⣿⣿⣶⣤⣤⣾⠛⠻⣷⡀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⣿⣿⡏⠉⠉⠙⠛⠿⠿⣷⣀⣀⣿⠃⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⡇⠀⠀⠀⠀⠀⠀⠈⠉⠉⠁⠀⠀
⠀⠀⣀⣤⣀⠀⢀⣠⣤⣶⣶⣿⣿⣿⣿⣿⣿⣶⣶⣤⣄⡀⠀⣀⣤⣀⠀⠀
⢰⡿⠋⢉⣹⣿⣿⣿⠿⠟⠛⠋⠉⠉⠉⠉⠙⠛⠻⠿⣿⣿⣿⣏⡉⠙⢿⡆
⢸⣇⣠⣾⣿⡿⠋⠀⠀⣠⣤⣀⠀⠀⠀⠀⣀⣤⣄⠀⠀⠙⢿⣿⣷⣄⣸⡗
⠈⢻⣿⣿⠋⠀⠀⠀⢸⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⡇⠀⠀⠀⠙⣿⣿⡟⠁
⠀⢸⣿⣿⠀⠀⠀⠀⠀⠉⠋⠁⠀⠀⠀⠀⠈⠙⠉⠀⠀⠀⠀⠀⣿⣿⡇⠀
⠀⠀⣿⣿⣧⡀⠀⠀⠀⢤⣀⡀⠀⠀⠀⠀⢀⣀⡤⠀⠀⠀⢀⣼⣿⣿⠀⠀
⠀⠀⠈⠿⣿⣷⣦⣀⠀⠀⠉⠻⠿⠿⠿⠿⠟⠉⠀⠀⣀⣴⣾⣿⠿⠁⠀⠀
⠀⠀⠀⠀⠉⠻⢿⣿⣿⣷⣶⣤⣤⣤⣤⣤⣤⣶⣾⣿⣿⡿⠟⠉⠀   Poly-REaDDIT: ⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⠉⠛⠛⠻⠿⠿⠿⠿⠟⠛⠛⠉⠁⠀⠀⠀⠀   a Reddit user info collector⠀⠀⠀
"""
        )

    def __init__(self, username):
        self.print_banner()
        self.reddit_session = requests.Session()
        self.reddit_session.headers = self.start_headers
        self.username = username
        self.get_reddit_user_info()
        self.get_NFT()

    def get_reddit_user_info(self):
        try:
            url = "https://www.reddit.com/user/" + self.username
            r = self.reddit_session.get(url)
            soup = BeautifulSoup(r.text, "html.parser")
            data = soup.find("script", {"id": "data"}).text
            body_data = json.loads(data.split(" = ")[1][:-1])
            created = datetime.utcfromtimestamp(
                body_data["users"]["models"][self.username.lower()]["createdUtc"]
            )
            self.created_str = created.strftime("%Y-%m-%d %H:%M:%S")
            self.user_id = body_data["users"]["models"][self.username.lower()]["id"]
            self.is_mod = body_data["users"]["models"][self.username.lower()]["isMod"]
            self.is_NSFW = body_data["users"]["models"][self.username.lower()]["isNSFW"]
            self.is_Gold = body_data["users"]["models"][self.username.lower()]["isGold"]
            self.is_Employee = body_data["users"]["models"][self.username.lower()][
                "isEmployee"
            ]
            self.is_Suspended = body_data["users"]["models"][self.username.lower()][
                "isSuspended"
            ]
            self.hide_from_robots = body_data["users"]["models"][self.username.lower()][
                "hideFromRobots"
            ]
            self.user_name = body_data["users"]["models"][self.username.lower()][
                "displayNamePrefixed"
            ]
            if body_data["users"]["models"][self.username.lower()]["hasUserProfile"]:
                self.profile_id = body_data["users"]["models"][self.username.lower()][
                    "profileId"
                ]
            self.user_karma = body_data["users"]["models"][self.username.lower()][
                "karma"
            ]["total"]
            self.user_karma_award_rec = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromAwardsReceived"]
            self.user_karma_award_given = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromAwardsGiven"]
            self.user_karma_comments = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromComments"]
            self.user_karma_post = body_data["users"]["models"][self.username.lower()][
                "karma"
            ]["fromPosts"]
            self.reddit_access_token = body_data["user"]["session"]["accessToken"]

        except:
            print(" [-] Error getting user info")
            return False

    def get_NFT(self):
        try:
            url = "https://gql.reddit.com/"
            self.session_api = requests.Session()
            _head = self.start_headers
            _head["Authorization"] = "Bearer " + self.reddit_access_token.replace(
                "'", ""
            )
            _head["Referer"] = "https://www.reddit.com/"
            _head["Origin"] = "https://www.reddit.com"
            _head["Content-Type"] = "application/json"
            self.session_api.headers = _head
            _payload = {
                "variables": {"id": self.user_id, "isLoggedInUserProfile": False},
                "id": "985a7e8ad0b9",
            }
            r = self.session_api.post(url, json=_payload)
            json_data = json.loads(r.text)
            for i in json_data["data"]["redditorInfoById"]["displayedCollectibleItems"][
                "edges"
            ]:
                nft = {}
                nft["name"] = i["node"]["item"]["name"]
                nft["id"] = i["node"]["item"]["id"]
                nft["serialNumber"] = i["node"]["item"]["serialNumber"]
                nft["title"] = i["node"]["item"]["nft"]["title"]
                nft["externalUrls"] = i["node"]["item"]["nft"]["externalUrls"][0]
                self.nfts.append(nft)
                if self.polygon_address == "N/A" and len(self.nfts) > 0:
                    self.get_token_owner(i["node"]["item"]["nft"]["externalUrls"][0])
        except:
            print(" [-] Error getting NFT")

    def print_data(self):
        print("")
        print(" [+] REDDIT USER INFO")
        print("  |")
        print("  ├ Username: {}".format(self.username))
        print("  ├ Created UTC: {}".format(self.created_str))
        print("  ├ User ID: {}".format(self.user_id))
        print("  ├ Is mod: {}".format(self.is_mod))
        print("  ├ Is NSFW: {}".format(self.is_NSFW))
        print("  ├ Is Gold: {}".format(self.is_Gold))
        print("  ├ Is Employee: {}".format(self.is_Employee))
        print("  ├ Is Suspended: {}".format(self.is_Suspended))
        print("  ├ Hide from robots: {}".format(self.hide_from_robots))
        print("  ├ Profile ID: {}".format(self.profile_id))
        print("  ├ User name: {}".format(self.user_name))
        print(
            "  ├ User karma: {} ({}: Awards received, {}: Awards given, {}: Comments, {}: Posts)".format(
                self.user_karma,
                self.user_karma_award_rec,
                self.user_karma_award_given,
                self.user_karma_comments,
                self.user_karma_post,
            )
        )
        print("  ├ Polygon Address: {}".format(self.polygon_address))
        print("  └ ------------------------------------")
        print("")

        if self.polygon_address != "N/A":
            print(" [+] NFT INFO")
            print("  |")
            for i in self.nfts:
                print("  ├ Name: {}".format(i["name"]))
                print("  ├ ID: {}".format(i["id"]))
                print("  ├ Serial Number: {}".format(i["serialNumber"]))
                print("  ├ Title: {}".format(i["title"]))
                print("  ├ External Urls: {}".format(i["externalUrls"]))
                print("  └ ------------------------------------")
                print("")
        else:
            print(" [-] No NFT INFO")
            print("")



    def get_token_owner(self, url):
        try:
            _query = url.split("token/")[1].split("?a=")
            _token = _query[0]
            _id = _query[1]
            text = requests.get(
                "https://opensea.io/assets/matic/{}/{}".format(_token, _id),
                headers=self.start_headers,
            ).text
            soup = BeautifulSoup(text, "html.parser")
            link = soup.find(
                "div", {"data-testid": "ItemOwnerAccountLink"}
            ).findChildren("a")
            self.polygon_address = link[0].attrs["href"].replace("/", "")
        except:
            print(" [-] Error getting token owner")

def run():
    parser = argparse.ArgumentParser()
    parser.add_argument("-username", help="reddit username")
    args = parser.parse_args()
    poly = PolyReaddit(args.username)
    poly.print_data()

if __name__ == "__main__":
    run()


src/polyreaddit/polyreaddit.py
import argparse
import requests
import json
from bs4 import BeautifulSoup
from datetime import datetime


class PolyReaddit:

    start_headers = {
        "Accept-Language": "us-US,us;q=0.8,en-US;q=0.5,en;q=0.3",
        "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/111.0",
    }
    session = None
    session_api = None

    username = ""
    created_str = ""
    nfts = []
    user_id = ""
    is_mod = ""
    is_NSFW = ""
    is_Gold = ""
    is_Employee = ""
    is_Suspended = ""
    hide_from_robots = ""
    profile_id = "N/A"
    user_name = ""
    user_karma = ""
    reddit_access_token = ""
    polygon_address = "N/A"
    user_messages = []

    def print_banner(self):
        print(
            """
⠀⠀⠀⢠⣾⣿⣿⣿⣿⣶⣤⣤⣾⠛⠻⣷⡀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⣿⣿⡏⠉⠉⠙⠛⠿⠿⣷⣀⣀⣿⠃⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⡇⠀⠀⠀⠀⠀⠀⠈⠉⠉⠁⠀⠀
⠀⠀⣀⣤⣀⠀⢀⣠⣤⣶⣶⣿⣿⣿⣿⣿⣿⣶⣶⣤⣄⡀⠀⣀⣤⣀⠀⠀
⢰⡿⠋⢉⣹⣿⣿⣿⠿⠟⠛⠋⠉⠉⠉⠉⠙⠛⠻⠿⣿⣿⣿⣏⡉⠙⢿⡆
⢸⣇⣠⣾⣿⡿⠋⠀⠀⣠⣤⣀⠀⠀⠀⠀⣀⣤⣄⠀⠀⠙⢿⣿⣷⣄⣸⡗
⠈⢻⣿⣿⠋⠀⠀⠀⢸⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⡇⠀⠀⠀⠙⣿⣿⡟⠁
⠀⢸⣿⣿⠀⠀⠀⠀⠀⠉⠋⠁⠀⠀⠀⠀⠈⠙⠉⠀⠀⠀⠀⠀⣿⣿⡇⠀
⠀⠀⣿⣿⣧⡀⠀⠀⠀⢤⣀⡀⠀⠀⠀⠀⢀⣀⡤⠀⠀⠀⢀⣼⣿⣿⠀⠀
⠀⠀⠈⠿⣿⣷⣦⣀⠀⠀⠉⠻⠿⠿⠿⠿⠟⠉⠀⠀⣀⣴⣾⣿⠿⠁⠀⠀
⠀⠀⠀⠀⠉⠻⢿⣿⣿⣷⣶⣤⣤⣤⣤⣤⣤⣶⣾⣿⣿⡿⠟⠉⠀   Poly-REaDDIT: ⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠈⠉⠛⠛⠻⠿⠿⠿⠿⠟⠛⠛⠉⠁⠀⠀⠀⠀   a Reddit user info collector⠀⠀⠀
"""
        )

    def __init__(self, username):
        self.print_banner()
        self.reddit_session = requests.Session()
        self.reddit_session.headers = self.start_headers
        self.username = username
        self.get_reddit_user_info()
        self.get_NFT()

    def get_reddit_user_info(self):
        try:
            url = "https://www.reddit.com/user/" + self.username
            r = self.reddit_session.get(url)
            soup = BeautifulSoup(r.text, "html.parser")
            data = soup.find("script", {"id": "data"}).text
            body_data = json.loads(data.split(" = ")[1][:-1])
            created = datetime.utcfromtimestamp(
                body_data["users"]["models"][self.username.lower()]["createdUtc"]
            )
            self.created_str = created.strftime("%Y-%m-%d %H:%M:%S")
            self.user_id = body_data["users"]["models"][self.username.lower()]["id"]
            self.is_mod = body_data["users"]["models"][self.username.lower()]["isMod"]
            self.is_NSFW = body_data["users"]["models"][self.username.lower()]["isNSFW"]
            self.is_Gold = body_data["users"]["models"][self.username.lower()]["isGold"]
            self.is_Employee = body_data["users"]["models"][self.username.lower()][
                "isEmployee"
            ]
            self.is_Suspended = body_data["users"]["models"][self.username.lower()][
                "isSuspended"
            ]
            self.hide_from_robots = body_data["users"]["models"][self.username.lower()][
                "hideFromRobots"
            ]
            self.user_name = body_data["users"]["models"][self.username.lower()][
                "displayNamePrefixed"
            ]
            if body_data["users"]["models"][self.username.lower()]["hasUserProfile"]:
                self.profile_id = body_data["users"]["models"][self.username.lower()][
                    "profileId"
                ]
            self.user_karma = body_data["users"]["models"][self.username.lower()][
                "karma"
            ]["total"]
            self.user_karma_award_rec = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromAwardsReceived"]
            self.user_karma_award_given = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromAwardsGiven"]
            self.user_karma_comments = body_data["users"]["models"][
                self.username.lower()
            ]["karma"]["fromComments"]
            self.user_karma_post = body_data["users"]["models"][self.username.lower()][
                "karma"
            ]["fromPosts"]
            self.reddit_access_token = body_data["user"]["session"]["accessToken"]
            self.retrieve_messages(body_data)
        except:
            print(" [-] Error getting user info")
            return False

    @staticmethod
    def get_last_timestamp(json_data, kind):
        if len(json_data[kind].keys()) > 0:
            return int(
                        json_data[kind][list(json_data[kind].keys())[len(json_data[kind].keys()) - 1]][
                            "created"])
        return 0


    def retrieve_messages(self, body_data):
        posts = True
        comments = True
        first_key = ""
        last_created = 0
        while True:
            if first_key == "":
                url = "https://gateway.reddit.com/desktopapi/v1/user/{}/conversations?rtj=only&redditWebClient=web2x&app=web2x-client-production&allow_over18=1&include=identity&dist=25&sort=new&t=all".format(
                    self.username)
            else:
                url = "https://gateway.reddit.com/desktopapi/v1/user/{}/conversations?rtj=only&redditWebClient=web2x&app=web2x-client-production&allow_over18=1&include=identity&after={}&dist=25&sort=new&t=all".format(
                    self.username, first_key)
            self.session_api = requests.Session()
            _head = self.start_headers
            _head["Authorization"] = "Bearer " + self.reddit_access_token.replace(
                "'", ""
            )
            preview_message = self.session_api.get(url, headers=_head).text
            json_message = json.loads(preview_message)

            if comments:
                for key in json_message["comments"].keys():
                    created = int(json_message["comments"][key]["created"])
                    message = { "created":created, "key": key, "permalink": json_message["comments"][key]["permalink"], "media": json_message["comments"][key]["media"] }
                    self.user_messages.append(message)
                if len(json_message["comments"].keys())>0:
                    timestamp_comment_created = PolyReaddit.get_last_timestamp(json_message, "comments")
                    if timestamp_comment_created > last_created:
                        last_created = timestamp_comment_created
                        first_key = list(json_message["comments"].keys())[len(json_message["comments"].keys()) - 1]

            if posts:
                for key in json_message["posts"].keys():
                    created = int(json_message["posts"][key]["created"])
                    message = { "created":created, "key": key, "permalink": json_message["posts"][key]["permalink"],
                               "media": json_message["posts"][key]["media"]}
                    self.user_messages.append(message)
                if len(json_message["posts"].keys())>0:
                    timestamp_post_created = PolyReaddit.get_last_timestamp(json_message, "posts")
                    if timestamp_post_created > last_created:
                        last_created = timestamp_post_created
                        first_key = list(json_message["posts"].keys())[len(json_message["posts"].keys()) - 1]

            if len(json_message["comments"].keys()) < 25:
                comments = False
            if len(json_message["posts"].keys()) < 25:
                posts = False

            if not posts and not comments:
                break
        print(" [+] Messages retrieved")


    def get_NFT(self):
        try:
            url = "https://gql.reddit.com/"
            self.session_api = requests.Session()
            _head = self.start_headers
            _head["Authorization"] = "Bearer " + self.reddit_access_token.replace(
                "'", ""
            )
            _head["Referer"] = "https://www.reddit.com/"
            _head["Origin"] = "https://www.reddit.com"
            _head["Content-Type"] = "application/json"
            self.session_api.headers = _head
            _payload = {
                "variables": {"id": self.user_id, "isLoggedInUserProfile": False},
                "id": "985a7e8ad0b9",
            }
            r = self.session_api.post(url, json=_payload)
            json_data = json.loads(r.text)
            for i in json_data["data"]["redditorInfoById"]["displayedCollectibleItems"][
                "edges"
            ]:
                nft = {}
                nft["name"] = i["node"]["item"]["name"]
                nft["id"] = i["node"]["item"]["id"]
                nft["serialNumber"] = i["node"]["item"]["serialNumber"]
                nft["title"] = i["node"]["item"]["nft"]["title"]
                nft["externalUrls"] = i["node"]["item"]["nft"]["externalUrls"][0]
                self.nfts.append(nft)
                if self.polygon_address == "N/A" and len(self.nfts) > 0:
                    self.get_token_owner(i["node"]["item"]["nft"]["externalUrls"][0])
        except:
            print(" [-] Error getting NFT")

    def print_data(self):
        print("")
        print(" [+] REDDIT USER INFO")
        print("  |")
        print("  ├ Username: {}".format(self.username))
        print("  ├ Created UTC: {}".format(self.created_str))
        print("  ├ User ID: {}".format(self.user_id))
        print("  ├ Is mod: {}".format(self.is_mod))
        print("  ├ Is NSFW: {}".format(self.is_NSFW))
        print("  ├ Is Gold: {}".format(self.is_Gold))
        print("  ├ Is Employee: {}".format(self.is_Employee))
        print("  ├ Is Suspended: {}".format(self.is_Suspended))
        print("  ├ Hide from robots: {}".format(self.hide_from_robots))
        print("  ├ Profile ID: {}".format(self.profile_id))
        print("  ├ User name: {}".format(self.user_name))
        print(
            "  ├ User karma: {} ({}: Awards received, {}: Awards given, {}: Comments, {}: Posts)".format(
                self.user_karma,
                self.user_karma_award_rec,
                self.user_karma_award_given,
                self.user_karma_comments,
                self.user_karma_post,
            )
        )
        print("  ├ Polygon Address: {}".format(self.polygon_address))
        print("  └ ------------------------------------")
        print("")

        if self.polygon_address != "N/A":
            print(" [+] NFT INFO")
            print("  |")
            for i in self.nfts:
                print("  ├ Name: {}".format(i["name"]))
                print("  ├ ID: {}".format(i["id"]))
                print("  ├ Serial Number: {}".format(i["serialNumber"]))
                print("  ├ Title: {}".format(i["title"]))
                print("  ├ External Urls: {}".format(i["externalUrls"]))
                print("  └ ------------------------------------")
                print("")
        else:
            print(" [-] No NFT INFO")
            print("")



    def get_token_owner(self, url):
        try:
            _query = url.split("token/")[1].split("?a=")
            _token = _query[0]
            _id = _query[1]
            text = requests.get(
                "https://opensea.io/assets/matic/{}/{}".format(_token, _id),
                headers=self.start_headers,
            ).text
            soup = BeautifulSoup(text, "html.parser")
            link = soup.find(
                "div", {"data-testid": "ItemOwnerAccountLink"}
            ).findChildren("a")
            self.polygon_address = link[0].attrs["href"].replace("/", "")
        except:
            print(" [-] Error getting token owner")

def run():
    parser = argparse.ArgumentParser()
    parser.add_argument("-username", help="reddit username")
    args = parser.parse_args()
    poly = PolyReaddit(args.username)
    poly.print_data()

if __name__ == "__main__":
    run()
